### PR TITLE
796: Run Requirements: don't allow duplicates

### PR DESCRIPTION
**Pull request by lostiniceland**
Requirements are checked with "equals" provided by BND. 
Ignored Requirements are displayed as information-message.

Signed-off-by: Marc Schlegel <maschlegel@gmail.com>

### CODE REVIEW

**Comment by bjhargrave** [comment]
Please review @njbartlett.

File: bndtools.core/src/bndtools/editor/project/RunRequirementsPart.java [hunk 1]
```diff
@@ -269,11 +271,7 @@ private void doAddBundle() {
                     Requirement req = new CapReqBuilder(IdentityNamespace.IDENTITY_NAMESPACE).addDirective(Namespace.REQUIREMENT_FILTER_DIRECTIVE, filter.toString()).buildSyntheticRequirement();
                     adding.add(req);
                 }
-                if (!adding.isEmpty()) {
```
Comment (bjhargrave) [inline bndtools.core/src/bndtools/editor/project/RunRequirementsPart.java]: Here you remove the empty check before calling the new method. Below you keep the empty check. I would assume these should be consistent.

File: bndtools.core/src/bndtools/editor/project/RunRequirementsPart.java [hunk 2]
```diff
@@ -474,9 +512,7 @@ protected boolean performSelectionDrop(ISelection data, Object target, int locat
             }
 
             if (!adding.isEmpty()) {
```
Comment (bjhargrave) [inline bndtools.core/src/bndtools/editor/project/RunRequirementsPart.java]: Here you leave the empty check before calling the new method. Above you remove the empty check. I would assume these should be consistent.

File: bndtools.core/src/bndtools/editor/project/RunRequirementsPart.java [hunk 3]
```diff
@@ -257,8 +259,8 @@ private void doAddBundle() {
             WizardDialog dialog = new WizardDialog(getSection().getShell(), wizard);
 
             if (Window.OK == dialog.open()) {
-                List<VersionedClause> result = wizard.getSelectedBundles();
-                List<Requirement> adding = new ArrayList<Requirement>(result.size());
+                final List<VersionedClause> result = wizard.getSelectedBundles();
+                final List<Requirement> adding = new ArrayList<Requirement>(result.size());
```
Comment (bjhargrave) [inline bndtools.core/src/bndtools/editor/project/RunRequirementsPart.java]: Couldn't this be much simpler if `adding` was a Set? Then duplicate adds would just be ignored.

File: bndtools.core/src/bndtools/editor/project/RunRequirementsPart.java (hunk 2 above)
Comment (lostiniceland) [inline bndtools.core/src/bndtools/editor/project/RunRequirementsPart.java]: This was on purpose. The intention was to remove the empty-check, but it was still necessary in the DropAdapter due to the boolean-return.

Going to add the check above for consistency.

Comment (bjhargrave) [inline bndtools.core/src/bndtools/editor/project/RunRequirementsPart.java]: Actually, I think the best solution is to simply replace the List with a Set. We don't need a dialog popup.

File: bndtools.core/src/bndtools/editor/project/RunRequirementsPart.java (hunk 3 above)
Comment (lostiniceland) [inline bndtools.core/src/bndtools/editor/project/RunRequirementsPart.java]: This depends on what you would like to display to the user:

If we dont want to display the duplicates to the user in the dialog, I can make this a Set. I decided to go for separate lists in order to give some feedback. Even though the dialog is not pretty and might anoy somebody I thought it to be better than a user who is wondering why nothing happened (editor marked dirty) after he added bundles.

Best solution might be to check if a bundle can be added during the selection but this has to be done in multiple places, so I went for the approach with the smallest impact.

Comment (bjhargrave) [inline bndtools.core/src/bndtools/editor/project/RunRequirementsPart.java]: Just do `adding.removeAll(requires)` to remove any existing requirements before the empty check. Then the editor will only be marked dirty if there are new requirements being added.

File: bndtools.core/src/bndtools/editor/project/RunRequirementsPart.java [hunk 4]
```diff
@@ -270,9 +272,7 @@ private void doAddBundle() {
                     adding.add(req);
                 }
                 if (!adding.isEmpty()) {
```
Comment (bjhargrave) [inline bndtools.core/src/bndtools/editor/project/RunRequirementsPart.java]: You do the adding check in the new method, so remove it here.

File: bndtools.core/src/bndtools/editor/project/RunRequirementsPart.java [hunk 5]
```diff
@@ -453,6 +453,23 @@ public void propertyChange(PropertyChangeEvent evt) {
...
     }
 
+    /**
+     * Update the requirements already available with new ones. Already existing requirements will be removed from the
+     * given set.
+     *
+     * @param adding
+     *            Set with {@link Requirement}s to add
+     */
+    private void updateViewerWithNewRequirements(Set<Requirement> adding) {
```
Comment (bjhargrave) [inline bndtools.core/src/bndtools/editor/project/RunRequirementsPart.java]: Have the method return boolean: !adding.isEmpty().

File: bndtools.core/src/bndtools/editor/project/RunRequirementsPart.java [hunk 6]
```diff
@@ -474,9 +491,7 @@ protected boolean performSelectionDrop(ISelection data, Object target, int locat
             }
 
             if (!adding.isEmpty()) {
```
Comment (bjhargrave) [inline bndtools.core/src/bndtools/editor/project/RunRequirementsPart.java]: Remove the isEmpty check and just return the result of the new method.

**Comment by bjhargrave** [comment]
Also please squash into a single commit. Thanks!

**Comment by bjhargrave** [comment]
I moved this work to https://github.com/bndtools/bndtools/pull/1253

**Pull request closed without merging**